libraries/AP_NMEA_Output/AP_NMEA_Output.cpp

```diff
@@ -163,7 +164,55 @@ void AP_NMEA_Output::update()
     char rmc_end[6];
     snprintf(rmc_end, sizeof(rmc_end), "*%02X\r\n", (unsigned) _nmea_checksum(rmc));
 
-    const uint32_t space_required = strlen(gga) + strlen(gga_end) + strlen(rmc) + strlen(rmc_end);
+
+    // get roll, pitch, yaw
+    const float roll_deg = wrap_180(degrees(ahrs.get_roll()));
+    const float pitch_deg = wrap_180(degrees(ahrs.get_pitch()));
+    const float yaw_deg = wrap_360(degrees(ahrs.get_yaw()));
+    const float heave_m = 0; // instantaneous heave in meters
+    const float roll_deg_accuracy = 0; // stddev of roll_deg;
+    const float pitch_deg_accuracy = 0; // stddev of pitch_deg;
+    const float heading_deg_accuracy = 0; // stddev of yaw_deg;
+
+    // GPS Update Quality Flag:
+    // 0 = no position
+    // 1 = All non-RTK fixed integer positions
+    // 2 = RTK fixed integer positions
+    const AP_GPS::GPS_Status gps_status = AP::gps().status();
+    const uint8_t gps_status_flag = (gps_status >= AP_GPS::GPS_OK_FIX_3D_RTK_FIXED) ? 2 :
+                                                (gps_status >= AP_GPS::GPS_OK_FIX_2D ? 1 : 0);
+
+    // INS Status Flag:
+    // 0 = All SPAN Pre-Alignment INS Status
+    // 1 = All SPAN Post-Alignment INS Status
+    const bool ins_status_flag = ahrs.initialised() &&
+                                    ahrs.healthy() &&
+                                    (!ahrs.have_inertial_nav() || AP::ins().accel_calibrated_ok_all());
+
+    // format PASHR message
+    char* pashr = nullptr;
+    int16_t pashr_res = asprintf(&pashr,
+                            "$PASHR,%s,%.2f,T,%c%.2f,%c%.2f,%c%.2f,%.3f,%.3f,%.3f,%u,%u",
+                            tstring,
+                            yaw_deg, // This is a TRUE NORTH value
+                            roll_deg<0? '-':'+', fabs(roll_deg),    // always show + or - symbol
+                            pitch_deg<0?'-':'+', fabs(pitch_deg),   // always show + or - symbol
+                            heave_m<0?  '-':'+', fabs(heave_m),     // always show + or - symbol
+                            roll_deg_accuracy,
+                            pitch_deg_accuracy,
+                            heading_deg_accuracy,
+                            (unsigned)gps_status_flag,
+                            (unsigned)ins_status_flag);
+    if (pashr_res == -1) {
+        free(gga);
```

Comment:
This free looks incorrect. I guess it should be pashr or we'd get a memory leak.  There's an [issue](https://github.com/ArduPilot/ardupilot/issues/12514) and [PR](https://github.com/ArduPilot/ardupilot/pull/22184) which suggests we shouldn't be allocating and freeing all the time.

Reply:
I agree with not constantly allocating, bad practice for our embedded stuff. Meanwhile, this frees the already allocated memory. If we get here then pashr_res is -1 which means it failed to allocate pashr. See example of GPRMC above this code

Reply:
Ah, ok, I see.